app/src/main/java/com/rajatgoyal/puzzle15/model/GameMatrix.java

```diff
@@ -0,0 +1,281 @@
+        int inv = getInversions();
+        int emptyCellRow = getEmptyCellRow();
+
+        return (emptyCellRow % 2 == 0 && inv % 2 != 0) || (emptyCellRow % 2 != 0 && inv % 2 == 0);
+    }
+
+    /**
+     * If puzzle is not solvable, make it solvable by decreasing one inversion
+     * which can be done easily by swapping two last positions
+     */
+    private void validateMatrix() {
+        int n1 = this.size - 1;
+        int n2 = this.size - 2;
+        int n3 = this.size - 3;
+
+        if (isValid()) {
+            return;
+        }
+
+        if (get(n1, n1) != 0) {
+            if (get(n1, n2) != 0) {
+                swap(n1, n2, n1, n1);
+            } else {
+                swap(n1, n3, n1, n1);
+            }
+        } else {
+            swap(n1, n3, n1, n2);
+        }
+    }
+
+    /**
+     * swap elements at given positions
+     *
+     * @param row1 element 1 row
+     * @param col1 element 1 column
+     * @param row2 element 2 row
+     * @param col2 element 2 column
+     */
+    public void swap(int row1, int col1, int row2, int col2) {
+        int temp = get(row1, col1);
+        set(row1, col1, get(row2, col2));
+        set(row2, col2, temp);
+    }
+
+    /**
+     * get element at given position
+     *
+     * @param row element row
+     * @param col element column
+     * @return element
+     */
+    public int get(int row, int col) {
+        return this.matrix[row][col];
+    }
+
+    /**
+     * set element at given position
+     *
+     * @param row   element row
+     * @param col   element column
+     * @param value element value
+     */
+    public void set(int row, int col, int value) {
+        this.matrix[row][col] = value;
+
+        if (value == 0) {
+            emptyCellRow = row;
+            emptyCellCol = col;
+        }
+    }
+
+    /**
+     * @return matrix is solved or not
+     */
+    public boolean isSolved() {
+        for (int i = 0; i < this.size; i++) {
+            for (int j = 0; j < this.size; j++) {
+                if (get(i, j) != this.size * i + j + 1)
+                    return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * @param formatted rows and columns are separated by different delimiters or by same one.
+     * @return game matrix string
+     */
+    public String toString(boolean formatted) {
+        StringBuilder stringBuilder = new StringBuilder();
+        if (formatted) {
+            for (int i = 0; i < this.size; ++i) {
+                if (i != 0) stringBuilder.append(ROW_SEPARATOR);
+                for (int j = 0; j < this.size; ++j) {
+                    if (j != 0) stringBuilder.append(COL_SEPARATOR);
+                    stringBuilder.append(get(i, j));
+                }
+            }
+            return stringBuilder.toString();
+        } else {
+            int[] array = getArray();
+            for (int i = 0; i < array.length; ++i) {
+                if (i != 0) stringBuilder.append(COL_SEPARATOR);
```

Comment:
Could have added first element as it is, and then start loop from `i = 1` and don't check for `i != 0` then.